### PR TITLE
 Block editor: Add `MediaUploadProvider` 

### PR DESCRIPTION
**Pull request by swissspidy**
> [!WARNING]
> **This is still work in progress and still incomplete!**

## What?

Adds a new `MediaUploadProvider` component to `block-editor` to leverage the new upload data store from the experimental `upload-media` package added in #66290.

To-do:

- [ ] Testing
- [ ] Ensure that this all still works if the experiment is disabled and that it doesn't interfere with npm publishing
- [ ] Update `getSettings().mediaUpload` in the `block-editor` package to use this new upload store (it should look like https://github.com/swissspidy/media-experiments/blob/8b64ed6150aca612cc649f32861c8049e79cd985/packages/editor/src/init/editor/media-upload.ts#L21-L109)

## Why?

Every block editor instance should have its own upload store instance

## How?

Follows the recommendation from https://github.com/WordPress/gutenberg/pull/66290#discussion_r1810383124 to implement `MediaUploadProvider` with `useSubRegistry`.

## Testing Instructions

Tests should still pass

Everything should still work with the experiment being disabled

### Testing Instructions for Keyboard

N/A

## Screenshots or screencast 

N/A

### CODE REVIEW

**Comment by swissspidy** [comment]
@youknowriad Is this roughly what you had in mind?

File: packages/block-editor/src/components/provider/index.js
Comment (youknowriad) [inline packages/block-editor/src/components/provider/index.js]: Nit: we could use a variable to hold the children to avoid "duplicating" the children.

File: packages/editor/README.md
Comment (youknowriad) [inline packages/editor/README.md]: not sure I understand why this is exported from the package.

File: packages/upload-media/src/components/provider/index.tsx
Comment (youknowriad) [inline packages/upload-media/src/components/provider/index.tsx]: I think updateSettings should be made a private selector (in block editor, it's a mistake that it's not private, this should only be edited by updating the settings prop basically)

**Review by youknowriad — COMMENTED** [review]
This looks great overall.

It's clear that's a bit too much added complexity but unfortunately, it's kind of necessary evil.

**Comment by youknowriad** [comment]
>  Update getSettings().mediaUpload in the block-editor package to use this new upload store (it should look like https://github.com/swissspidy/media-experiments/blob/8b64ed6150aca612cc649f32861c8049e79cd985/packages/editor/src/init/editor/media-upload.ts#L21-L109)

Correct me if I'm wrong but for me the setting passed to the block editor, shouldn't use any store. It's a setting that will be used internally by the new package to actually perform the upload right?

File: packages/editor/README.md
Comment (swissspidy) [reply packages/editor/README.md]: Right, there is no need.

**Comment by swissspidy** [comment]
Yes. What I mean by this is:

The function passed to the block editor will be passed on unmodified to the upload store for eventually uploading the file to the server.

Now, if someone, for example the image block, calls `select( blockEditorStore ).getSettings().mediaUpload()`, this is a **different function** that will do some light validation and then pass the file to the uploadStore for further processing and eventual uploading to the server (using the original function provided).

I just need to figure out how to best do that :)

File: packages/upload-media/src/components/provider/index.tsx
Comment (swissspidy) [reply packages/upload-media/src/components/provider/index.tsx]: Will change this in the parent PR and then update here.

**Review by youknowriad — APPROVED** [review]
So this is looking good to me overall. Let me know if you want to merge this into the other PR or if we should review the other PR separate from this to help some people looking at the diff maybe. No preference for me.

**Comment by swissspidy** [comment]
Awesome, thanks! Yeah let's review the other PR first to make things easier :) Appreciate any eyes there.

File: packages/upload-media/src/components/provider/with-registry-provider.tsx
Comment (sirreal) [inline packages/upload-media/src/components/provider/with-registry-provider.tsx]: The build-types directory should not be referenced directly. I think the registry type could probably just be exported from the data package, it's effectively public already.

If we don't want to export it here, it can be accessed in a better way like this:

```suggestion
import { createHigherOrderComponent } from '@wordpress/compose';

/**
 * Internal dependencies
 */
import { storeConfig } from '../../store';
import { STORE_NAME as mediaUploadStoreName } from '../../store/constants';

type WPDataRegistry = ReturnType< typeof createRegistry >;

```

**Pull request merged**
